From 6b22dab7773eb55719e9cd7c04795d07187b651a Mon Sep 17 00:00:00 2001
From: Sudharsan Eswaran <eswaran@qti.qualcomm.com>
Date: Tue, 2 Jun 2026 14:49:03 +0530
Subject: [PATCH] Data model update on receiving channel preference report
 message

1. modified channel preference data model as per TR181 standars -
   Device.WiFi.DataElements.Network.Device.{i}.Radio.{i}.OpClassPreference.{i}
2. Added OpclassPreference event to intimate the preference has update to the
   application dependent on datamodel

Signed-off-by: Sudharsan Eswaran <eswaran@qti.qualcomm.com>
---
 controller/nbapi/odl/controller.odl.in        |  4 ++
 controller/nbapi/odl/radio.odl.in             | 13 ++--
 controller/src/beerocks/master/db/db.cpp      | 62 ++++++++++++++-----
 controller/src/beerocks/master/db/db.h        |  3 +
 .../dynamic_channel_selection_r2_task.cpp     |  2 +
 5 files changed, 58 insertions(+), 26 deletions(-)

diff --git a/controller/nbapi/odl/controller.odl.in b/controller/nbapi/odl/controller.odl.in
index fc1a3f2eb..07d2c8a64 100644
--- a/controller/nbapi/odl/controller.odl.in
+++ b/controller/nbapi/odl/controller.odl.in
@@ -605,6 +605,10 @@
                         /* Vendor event: raised when a beacon measurement report is received for a STA */
                         object BeaconMeasurement {
                         }
+
+                        /* Vendor event: raised when channel preference report received */
+                        object OpClassPreference {
+                        }
                     }
 
                     /* Update all unassociated stations stats, from all agents and all radios */
diff --git a/controller/nbapi/odl/radio.odl.in b/controller/nbapi/odl/radio.odl.in
index e7a9ac8ec..389f0ce5d 100644
--- a/controller/nbapi/odl/radio.odl.in
+++ b/controller/nbapi/odl/radio.odl.in
@@ -315,10 +315,10 @@
         }
 
         /* Channel preference reported by the agent for this radio */
-        %read-only object ChannelPreference [] {
-            counted with ChannelPreferenceNumberOfEntries;
+        %read-only object OpClassPreference [] {
+            counted with OpClassPreferenceNumberOfEntries;
 
-            %read-only uint32 OperatingClass = 0 {
+            %read-only uint32 OpClass = 0 {
                 on action validate call check_range [0, 255];
             }
             %read-only uint32 Preference = 0 {
@@ -328,12 +328,7 @@
                 on action validate call check_range [0, 15];
             }
 
-            %read-only object Channel [] {
-                counted with ChannelNumberOfEntries;
-                %read-only uint32 Channel = 0 {
-                    on action validate call check_range [0, 255];
-                }
-            }
+            %read-only string ChannelList = "";
         }
 
     }
diff --git a/controller/src/beerocks/master/db/db.cpp b/controller/src/beerocks/master/db/db.cpp
index 654fa002f..ca90eb6dd 100644
--- a/controller/src/beerocks/master/db/db.cpp
+++ b/controller/src/beerocks/master/db/db.cpp
@@ -3784,7 +3784,7 @@ bool db::dm_set_channel_preference(
         return true;
     }
 
-    const std::string channel_preference_path = radio->dm_path + ".ChannelPreference";
+    const std::string channel_preference_path = radio->dm_path + ".OpClassPreference";
 
     // Clear existing ChannelPreference instances before repopulating
     if (!m_ambiorix_datamodel->remove_all_instances(channel_preference_path)) {
@@ -3812,7 +3812,7 @@ bool db::dm_set_channel_preference(
         }
 
         bool ret_val = true;
-        ret_val &= m_ambiorix_datamodel->set(instance_path, "OperatingClass", op_cls_num);
+        ret_val &= m_ambiorix_datamodel->set(instance_path, "OpClass", op_cls_num);
         ret_val &= m_ambiorix_datamodel->set(instance_path, "Preference", flags.preference);
         ret_val &= m_ambiorix_datamodel->set(instance_path, "ReasonCode", flags.reason_code);
 
@@ -3823,8 +3823,8 @@ bool db::dm_set_channel_preference(
         }
 
         // Populate the Channel sub-list for this operating class
-        const auto channel_list_path = instance_path + ".Channel";
         const auto channel_count     = operating_class.channel_list_length();
+        std::string channel_list = "";
 
         for (int ch_idx = 0; ch_idx < channel_count; ch_idx++) {
             auto channel_ptr = operating_class.channel_list(ch_idx);
@@ -3833,25 +3833,25 @@ bool db::dm_set_channel_preference(
                            << int(op_cls_num) << " on radio " << radio_uid;
                 continue;
             }
-
-            auto channel_instance = m_ambiorix_datamodel->add_instance(channel_list_path);
-            if (channel_instance.empty()) {
-                LOG(ERROR) << "Failed to add Channel instance under " << channel_list_path
-                           << " for operating class " << int(op_cls_num) << " on radio "
-                           << radio_uid;
-                return false;
+            if (!channel_list.empty()) {
+                channel_list += ",";
             }
+            channel_list += std::to_string(static_cast<uint32_t>(*channel_ptr));
+        }
 
-            if (!m_ambiorix_datamodel->set(channel_instance, "Channel",
-                                           static_cast<uint32_t>(*channel_ptr))) {
-                LOG(ERROR) << "Failed to set Channel value " << int(*channel_ptr)
-                           << " for operating class " << int(op_cls_num) << " on radio "
-                           << radio_uid;
-                return false;
-            }
+        if (!m_ambiorix_datamodel->set(instance_path, "ChannelList", channel_list)) {
+            LOG(ERROR) << "Failed to set ChannelList (" << channel_list << ") for operating class "
+                       << int(op_cls_num) << " on radio " << radio_uid;
+            return false;
         }
     }
 
+    // Trigger Channel Preference Report QCOM vendor event after parsing
+    auto agent = get_agent_by_radio_uid(radio_uid);
+    if (agent) {
+        dm_trigger_channel_preference_report_event(radio_uid, radio->dm_path, agent->al_mac);
+    }
+
     return true;
 }
 
@@ -10246,6 +10246,34 @@ std::shared_ptr<Agent::sEthSwitch> db::get_eth_switch(const sMacAddr &mac)
 }
 
 #ifdef ENABLE_NBAPI
+bool db::dm_trigger_channel_preference_report_event(const sMacAddr &radio_mac, const std::string &radio_dm_path, const sMacAddr &agent_al_mac)
+{
+    amxc_var_t event_data;
+    amxc_var_init(&event_data);
+    amxc_var_set_type(&event_data, AMXC_VAR_ID_HTABLE);
+
+    const std::string radio_mac_str = tlvf::mac_to_string(radio_mac);
+    const std::string al_mac_str = tlvf::mac_to_string(agent_al_mac);
+
+    amxc_var_add_key(cstring_t, &event_data, "MACAddress", radio_mac_str.c_str());
+    amxc_var_add_key(cstring_t, &event_data, "DMPath", radio_dm_path.c_str());
+    amxc_var_add_key(cstring_t, &event_data, "ALMAC", al_mac_str.c_str());
+
+    const std::string vendor_event_obj = DATAELEMENTS_ROOT_DM ".Network.QCOM_Vendor.OpClassPreference";
+    const bool ok = m_ambiorix_datamodel->trigger_event_with_data(vendor_event_obj, "Event", &event_data);
+
+    if (!ok) {
+        LOG(ERROR) << "Failed to trigger OpClassPreference at " << vendor_event_obj
+                   << " radio_mac=" << radio_mac_str << " dm_path=" << radio_dm_path
+                   << " al_mac=" << al_mac_str;
+    } else {
+        LOG(DEBUG) << "Triggered OpClassPreference radio_mac=" << radio_mac_str
+                   << " dm_path=" << radio_dm_path << " al_mac=" << al_mac_str;
+    }
+    amxc_var_clean(&event_data);
+    return ok;
+}
+
 bool db::dm_trigger_qcom_vendor_event(const sMacAddr &mac, const std::string &dm_path,
                                       const std::string &event_name)
 {
diff --git a/controller/src/beerocks/master/db/db.h b/controller/src/beerocks/master/db/db.h
index 234743b9c..82840900e 100644
--- a/controller/src/beerocks/master/db/db.h
+++ b/controller/src/beerocks/master/db/db.h
@@ -2545,6 +2545,9 @@ public:
     bool dm_trigger_operating_channel_report_event(const sMacAddr &radio_mac,
                                                    const std::string &radio_dm_path,
                                                    const sMacAddr &agent_al_mac);
+    bool dm_trigger_channel_preference_report_event(const sMacAddr &radio_mac,
+                                                    const std::string &radio_dm_path,
+                                                    const sMacAddr &agent_al_mac);
     bool dm_trigger_assoc_sta_extended_link_metrics_event(const sMacAddr &sta_mac,
                                                           const std::string &sta_dm_path,
                                                           const sMacAddr &bss_mac,
diff --git a/controller/src/beerocks/master/tasks/dynamic_channel_selection_r2_task.cpp b/controller/src/beerocks/master/tasks/dynamic_channel_selection_r2_task.cpp
index 1645a5403..6547f55c8 100644
--- a/controller/src/beerocks/master/tasks/dynamic_channel_selection_r2_task.cpp
+++ b/controller/src/beerocks/master/tasks/dynamic_channel_selection_r2_task.cpp
@@ -1796,6 +1796,8 @@ bool dynamic_channel_selection_r2_task::handle_tlv_channel_preference(
     }
 
     LOG(INFO) << ss.str();
+
+    database.dm_set_channel_preference(channel_preference_tlv);
     return true;
 }
 
-- 
2.34.1

